plugins/src/main/java/com/google/fhir/gateway/plugin/PatientAccessChecker.java

```diff
@@ -168,24 +242,116 @@ private AccessDecision processBundle(RequestDetailsReader requestDetails) {
         return NoOpAccessDecision.accessDenied();
       }
     }
+
+    Bundle requestBundle = createBundleFromRequest(requestDetails);
+    if (requestBundle == null) {
+      return NoOpAccessDecision.accessDenied();
+    }
+    for (BundleEntryComponent entryComponent : requestBundle.getEntry()) {
+      if (!doesBundleElementHavePermission(entryComponent)) {
+        return NoOpAccessDecision.accessDenied();
+      }
+    }
     return NoOpAccessDecision.accessGranted();
   }
 
+  private Bundle createBundleFromRequest(RequestDetailsReader request) {
```

Comment:
Two duplication problems here:
- Performance: We are parsing the `Bundle` twice, once here and once in the `patientFinder.findPatientsInBundle` call above.
- Code: Parts of the code here are similar to `PatientFinderImp`.

Code duplication is easy to fix (we should just refactor the common code). For the performance issue, I am not sure what is best. We can expand the interface of `PatientFinder` to return more information about individual entries (but I don't like it as it is very SoF specific).

A better option is to pass a callback interface to `PatientFinder.indPatientsInBundle` such that for each entry, that callback is called and makes a decision whether the access should still be granted (the callback receives a `BundleEntryComponent` and returns true/false). So we leave the Bundle parsing to PatientFinder (in a single place and done only once per request) but for each entry, we check the callback too. WDYT?

Reply:
I did think of this as well. `PatientFinder` by its name and functionality is supposed to do very patient specific operations on a bundle.  In this case  if we add a callback, we would need to store the callback responses in `BundlePatients` , we do end up expanding the `BundlePatients` scope to access decisions. 

Other option I see is to create a `BundleProcessor` separately from the `PatientFinder` and move all the bundle processing logic from the `PatientFinder` to this new entity. In this new class, we can chain multiple operations on a `BundleEntryComponent` based on callbacks like finding patient IDs, checking access based on resource permissions etc. We can provide implementations of some common callbacks similar to  PatientFinder.

Reply:
Yes, I like your approach, so `PatientFinder` would be one consumer of `BundleProcessor` and our SoF use-case here would be another one. Another benefit of that approach is that this is a common pattern for any access-checker, where we need to implement `processGet`, `processPut`, etc. individually and also do the same for Bundle entries. So by encapsulating Bundle processing in a separate class we make all access-checker implementations simpler.

If you like to do that in this PR that is fine. If you like to finish this PR with your current approach (which we know has duplications) and follow-up with another PR immediately after it to do the refactoring, that is fine too; so up to you. In the latter case, please create an issue describing what the refactoring is and reference it to this PR as a TODO. Note that we should also change ListAccessChecker as it has a similar pattern as well.

Reply:
Filed an issue: #104 , will follow this up with a separate PR